clap_complete/src/shells/zsh.rs

```diff
@@ -622,18 +622,49 @@ fn write_positionals_of(p: &Command) -> String {
 
     let mut ret = vec![];
 
-    for arg in p.get_positionals() {
+    // The index of the 2nd-to-last positional argument.
+    let positionals: Vec<_> = p.get_positionals().collect();
+    let penultimate_idx = if positionals.len() > 1 {
+        Some(positionals.len() - 2)
+    } else {
+        None
+    };
+    // Completions for commands that end with two Vec arguments require special care:
+    // - You can have two Vec args separated with a value_terminator.
+    // - You can have two Vec args with the second one set to last (raw sets last)
+    //   which will force a separator of -- before the second.
+    // zsh requires a terminator in both of these situations.
+    let needs_terminator = match positionals.last() {
+        Some(arg) => arg.is_last_set() || arg.get_num_args().expect("built").max_values() > 1,
+        None => false,
+    };
+
+    for (idx, arg) in positionals.iter().enumerate() {
         debug!("write_positionals_of:iter: arg={}", arg.get_id());
 
         let num_args = arg.get_num_args().expect("built");
         let cardinality_value;
         let cardinality = if num_args.max_values() > 1 {
+            // The current argument takes multiple values.
+            // Is it the penultimate argument?
+            let is_penultimate = penultimate_idx.map(|value| value == idx).unwrap_or(false);
             match arg.get_value_terminator() {
                 Some(terminator) => {
                     cardinality_value = format!("*{}:", escape_value(terminator));
                     cardinality_value.as_str()
                 }
-                None => "*:",
+                None => {
+                    // If this is the penultimate argument (which takes multiple
+                    // arguments), and the next (final) argument also takes multiple
+                    // values, then a terminator must be used to separate the arguments.
+                    // "--" is used when no value terminator has been specified.
+                    // Otherwise, no terminator is needed and we can use "*".
+                    if is_penultimate && needs_terminator {
```

Comment:
While this works for a subset of cases, when an arg is marked `last`, it always needs a `--` before it.  Is there a better way to handle that as this feels a bit hacky to put it on the argument before.

For example, we should be able to have an arg that has a `value_terminator` **and** have a `last`.

Reply:
Agreed. I believe the idea in the previous comment may cover this scenario if it pans out as expected.